module/zfs/dmu.c

```diff
@@ -499,10 +502,12 @@ dmu_buf_hold_array_by_dnode(dnode_t *dn, uint64_t offset, uint64_t length,
 	rw_exit(&dn->dn_struct_rwlock);
 
 	/* wait for async i/o */
-	err = zio_wait(zio);
-	if (err) {
-		dmu_buf_rele_array(dbp, nblks, tag);
-		return (err);
+	if (zio) {
+		err = zio_wait(zio);
+		if (err) {
+			dmu_buf_rele_array(dbp, nblks, tag);
+			return (err);
+		}
```

Comment:
It looks like this whole block could be pulled under the existing read conditional below.

Reply:
@behlendorf thanks for the review, please have a look at slighly updated version - basically it helps reads too. this doesn't address your concerns yet, but I'm fine to fix those of course, if the approach taken in the patch makes some sense.

Reply:
The write side looks good.  But I'm fairly certain we still need to call `dbuf_read` even when the `db_state == DB_CACHED` since it's possible the cached dbuf still needs to be decompressed or decrypted.  Since you saw a significant performance win in your testing by skipping `DB_CACHED` dbufs it'd be interesting to know which bit of the `db->db_state == DB_CACHED` case in `dbuf_read()` was slowing things down.  Perhaps that can be optimized.